### PR TITLE
g.extension: match module files on Windows

### PR DESCRIPTION
**Pull request by ninsbl**
Just a swift attempt to fix some issues with installation of extensions on windows.
Cuts number of failing addons to half (~) compared to:
https://gist.github.com/ninsbl/8051f51d372b565ca6f5a29b480d2463

There you also find a test-recipie...

The remaining need to be fixed on AddOn side, I guess...

### CODE REVIEW

**Comment by ninsbl** [comment]
Adress #1499
Warnings for modules remain, but libs are removed...

**Comment by hellik** [comment]
g.extension locally patched:

```
g.extension extension=i.sentinel                                                
Downloading precompiled GRASS Addons <i.sentinel>...
Fetching <i.sentinel> from <http://wingrass.fsv.cvut.cz/grass79/x86_64/addons/grass-7.9.dev/i.sentinel.zip> (be patient)...
Updating extensions metadata file...
Updating extension modules metadata file...
WARNING: No metadata available for module 'i.sentinel.download': Unable to fetch interface description for command '<i.sentinel.download>'.

Details: <Cannot find the executable i.sentinel.download>
WARNING: No metadata available for module 'i.sentinel': Unable to fetch interface description for command '<i.sentinel>'.

Details: <Cannot find the executable i.sentinel>
ERROR: Unable to read manual page: [Errno 2] No such file or directory: 'C:\\Users\\user\\AppData\\Roaming\\GRASS7\\addons\\docs\\html\\i.sentinel.download.html'
```

it seems i.sentinel.download.html isn't build on the addon server.

original message was:

```
g.extension extension=i.sentinel                                                
Downloading precompiled GRASS Addons <i.sentinel>...
Fetching <i.sentinel> from <http://wingrass.fsv.cvut.cz/grass78/x86_64/addons/grass-7.8.5/i.sentinel.zip> (be patient)...
Updating extensions metadata file...
Updating extension modules metadata file...
WARNING: No metadata available for module 'i.sentinel.download'.
ERROR: Unable to read manual page: [Errno 2] No such file or directory: 'C:\\Users\\youruser\\AppData\\Roaming\\GRASS7\\addons\\docs\\html\\i.sentinel.download.html'
```

**Comment by hellik** [comment]
adding the environmental variable:

set PYTHONUTF8=1

```
g.extension extension=i.fusion.hpf                                              
Downloading precompiled GRASS Addons <i.fusion.hpf>...
Fetching <i.fusion.hpf> from <http://wingrass.fsv.cvut.cz/grass79/x86_64/addons/grass-7.9.dev/i.fusion.hpf.zip> (be patient)...
Updating extensions metadata file...
Updating extension modules metadata file...
Installation of <i.fusion.hpf> successfully finished
```
this looks good now; original message:

```
g.extension extension=i.fusion.hpf                                              
Downloading precompiled GRASS Addons <i.fusion.hpf>...
Fetching <i.fusion.hpf> from <http://wingrass.fsv.cvut.cz/grass79/x86_64/addons/grass-7.9.dev/i.fusion.hpf.zip> (be patient)...
Updating extensions metadata file...
Updating extension modules metadata file...
WARNING: No metadata available for module 'constants'.
WARNING: No metadata available for module 'high_pass_filter'.
ERROR: Unable to read manual page: [Errno 2] No such file or directory: 'C:\\Users\\youruser\\AppData\\Roaming\\GRASS7\\addons\\docs\\html\\constants.html'
```

**Comment by ninsbl** [comment]
> it seems i.sentinel.download.html isn't build on the addon server.

Yes, I am on it. It is the `requests` import causing trouble, as `requests` is a non-standard module...

**Comment by hellik** [comment]
> 
> 
> > it seems i.sentinel.download.html isn't build on the addon server.
> 
> Yes, I am on it. It is the `requests` import causing trouble, as `requests` is a non-standard module...

ok, a lazy import needed?

**Review by hellik — COMMENTED** [review]
looks better now

**Comment by hellik** [comment]
with

`if re.search(r"^[d,db,g,i,m,r,r3,v]\..*[\.py,\.exe]$", file):`

```
g.extension extension=i.sentinel                                                
Downloading precompiled GRASS Addons <i.sentinel>...
Fetching <i.sentinel> from <http://wingrass.fsv.cvut.cz/grass79/x86_64/addons/grass-7.9.dev/i.sentinel.zip> (be patient)...
Updating extensions metadata file...
Updating extension modules metadata file...
WARNING: No metadata available for module 'i.sentinel.download': Unable to fetch interface description for command '<i.sentinel.download>'.

Details: <Cannot find the executable i.sentinel.download>
ERROR: Unable to read manual page: [Errno 2] No such file or directory: 'C:\\Users\\hkmyr\\AppData\\Roaming\\GRASS7\\addons\\docs\\html\\i.sentinel.download.html'
```

**Comment by ninsbl** [comment]
> ok, a lazy import needed?

Yes, there are a couple of other modules with similar issues. Preparing a PR in addons...

**Comment by hellik** [comment]
> 
> 
> > ok, a lazy import needed?
> 
> Yes, there are a couple of other modules with similar issues. Preparing a PR in addons...

thanks for fixing these.

**Comment by hellik** [comment]
> 
> 
> > with
> > `if re.search(r"^[d,db,g,i,m,r,r3,v]\..*[\.py,\.exe]$", file):`
> 
> Hmmmm... For me (with GRASS 7.8.5 from OSGeo4W next generation), results with and without `$`are identical....
> No idea. Maybe it is safe also without $...

testing with winGRASS 7.8.5 OSGeo4W next generation (python 3.9)

`if re.search(r"^[d,db,g,i,m,r,r3,v]\..*[\.py,\.exe]$", file):`

```
Fetching <i.sentinel> from <http://wingrass.fsv.cvut.cz/grass78/x86_64/addons/grass-7.8.5/i.sentinel.zip> (be patient)...
Updating extensions metadata file...
Updating extension modules metadata file...
WARNING: No metadata available for module 'i.sentinel.download'.
ERROR: Unable to read manual page: [Errno 2] No such file or directory: 'C:\\Users\\hkmyr\\AppData\\Roaming\\GRASS7\\addons\\docs\\html\\i.sentinel.download.html'
```
tested with

`if re.search(r"^[d,db,g,i,m,r,r3,v]\..*[\.py,\.exe]", file):`

```
g.extension extension=i.sentinel                                                
Downloading precompiled GRASS Addons <i.sentinel>...
Fetching <i.sentinel> from <http://wingrass.fsv.cvut.cz/grass78/x86_64/addons/grass-7.8.5/i.sentinel.zip> (be patient)...
Updating extensions metadata file...
Updating extension modules metadata file...
WARNING: No metadata available for module 'i.sentinel.download'.
WARNING: No metadata available for module 'i.sentinel'.
ERROR: Unable to read manual page: [Errno 2] No such file or directory: 'C:\\Users\\hkmyr\\AppData\\Roaming\\GRASS7\\addons\\docs\\html\\i.sentinel.download.html'
```

there is a difference with `$` (see above)

**Review by hellik — APPROVED** [review]

**Comment by ninsbl** [comment]
Still not 100% sure about the `$` issue. `i.sentinel`should not be recognized as a module, so, the messages with `$` seem more correct. But remove my addons on Windows and retry installation to see if some are missing...

**Comment by ninsbl** [comment]
There was an issue with temporal modules. They were not included in the filter and thus not installed. I checked som manually.
`t.rast.out.xyz` fails now:, did it fail before?
```g.extension extension=t.rast.out.xyz
Downloading precompiled GRASS Addons <t.rast.out.xyz>...
Fetching <t.rast.out.xyz> from
<http://wingrass.fsv.cvut.cz/grass78/x86_64/addons/grass-7.8.5/t.rast.out.xyz.zip>
(be patient)...
Updating extensions metadata file...
Updating extension modules metadata file...
WARNING: No metadata available for module 't.rast.out.xyz': Unable to fetch
         interface description for command '<t.rast.out.xyz>'.

         Details: <C:\OSGEO4~2\bin\python3.exe: can't find '__main__'
         module in
         'C:\\Users\\STEFAN~1.BLU\\AppData\\Local\\Temp\\grass7-stefan.blumentrath-40776\\tmp_1ik4too\\t.rast.out.xyz'
         >
Installation of <t.rast.out.xyz> successfully finished
```

`t.rast.what.aggr` installs without issues...

Did `t.rast.null` or `t.rast.kappa` fail on Windows also before or is this an issue with this PR?

**Comment by hellik** [comment]
> Still not 100% sure about the `$` issue. `i.sentinel`should not be recognized as a module, so, the messages with `$` seem more correct. But remove my addons on Windows and retry installation to see if some are missing...

See my comments in https://github.com/OSGeo/grass/pull/1565#issuecomment-834722569

**Comment by hellik** [comment]
> There was an issue with temporal modules. They were not included in the filter and thus not installed. I checked som manually.
> `t.rast.out.xyz` fails now:, did it fail before?
> 
> ```
> Downloading precompiled GRASS Addons <t.rast.out.xyz>...
> Fetching <t.rast.out.xyz> from
> <http://wingrass.fsv.cvut.cz/grass78/x86_64/addons/grass-7.8.5/t.rast.out.xyz.zip>
> (be patient)...
> Updating extensions metadata file...
> Updating extension modules metadata file...
> WARNING: No metadata available for module 't.rast.out.xyz': Unable to fetch
>          interface description for command '<t.rast.out.xyz>'.
> 
>          Details: <C:\OSGEO4~2\bin\python3.exe: can't find '__main__'
>          module in
>          'C:\\Users\\STEFAN~1.BLU\\AppData\\Local\\Temp\\grass7-stefan.blumentrath-40776\\tmp_1ik4too\\t.rast.out.xyz'
>          >
> Installation of <t.rast.out.xyz> successfully finished
> ```
> 
> `t.rast.what.aggr` installs without issues...
> 
> Did `t.rast.null` or `t.rast.kappa` fail on Windows also before or is this an issue with this PR?

Never used them before, so I can't tell anything if it worked

**Comment by ninsbl** [comment]
OK, without the PR, t.rast.kappa and t.rast.null install nicely also on Windows. So, that is a new issue introduced with this PR that needs to be fixed...

**Comment by nilason** [comment]
> OK, without the PR, t.rast.kappa and t.rast.null install nicely also on Windows. So, that is a new issue introduced with this PR that needs to be fixed...

 `t.rast.kappa`, `t.rast.null` and `t.rast.out.xyz` installed without issues, **with** this PR and 7.9.dev. Don't know about 7.8.

`i.sentinel.download` probably fails on Win dist because of missing python module `request`. Try `import request` in Python Console.

**Comment by hellik** [comment]
> 
> 
> > `i.sentinel.download` probably fails on Win dist because of missing python module `request`. Try `import request` in Python Console.
> 
> I should add that a `bin\i.sentinel.download.bat` file is missing from addon zipfile retrieved from `http://wingrass.fsv.cvut.cz/grass78/x86_64/addons/grass-7.8.5/i.sentinel.zip`. The reason for this could possibly be related to the missing python module.

there is a [PR](https://github.com/OSGeo/grass-addons/pull/524) about lazy import of request that addon can be build on the server.

**Review by nilason — APPROVED** [review]
This looks good to me. Probably solves most – if not all – encoding problems on Windows installing addons with g.extension. Should be backported.

**Comment by ninsbl** [comment]
> There was an issue with temporal modules. They were not included in the filter and thus not installed. I checked som manually.
> `t.rast.out.xyz` fails now:, did it fail before?
> 
> ```
> Downloading precompiled GRASS Addons <t.rast.out.xyz>...
> Fetching <t.rast.out.xyz> from
> <http://wingrass.fsv.cvut.cz/grass78/x86_64/addons/grass-7.8.5/t.rast.out.xyz.zip>
> (be patient)...
> Updating extensions metadata file...
> Updating extension modules metadata file...
> WARNING: No metadata available for module 't.rast.out.xyz': Unable to fetch
>          interface description for command '<t.rast.out.xyz>'.
> 
>          Details: <C:\OSGEO4~2\bin\python3.exe: can't find '__main__'
>          module in
>          'C:\\Users\\STEFAN~1.BLU\\AppData\\Local\\Temp\\grass7-stefan.blumentrath-40776\\tmp_1ik4too\\t.rast.out.xyz'
>          >
> Installation of <t.rast.out.xyz> successfully finished
> ```
> 
> `t.rast.what.aggr` installs without issues...
> 
> Did `t.rast.null` or `t.rast.kappa` fail on Windows also before or is this an issue with this PR?



> OK, without the PR, t.rast.kappa and t.rast.null install nicely also on Windows. So, that is a new issue introduced with this PR that needs to be fixed...

My bad. I manually copied the modified version of g.extension from master into my MS Windows installation. 

That was the reason for a different warning message. t.rast.kappa installs just fine.

And when I tried to run it it failed due to lack of ctypes loading, but that is probably an issue with next generation OSGeo4W I had to use as that does not require admin access...

So, in short, should be OK to merge...

**Comment by neteler** [comment]
Shall I backport it right away or do you prefer to wait?

**Comment by nilason** [comment]
> Shall I backport it right away or do you prefer to wait?

I’d say no need to wait. On the contrary, the sooner the better. Needs to be tested before final 7.8.6 release.

**Comment by neteler** [comment]
> > Shall I backport it right away or do you prefer to wait?
> 
> I’d say no need to wait. On the contrary, the sooner the better. Needs to be tested before final 7.8.6 release.

Backport done!

**Comment by hellik** [comment]
thanks; related issue #1499 closed
